### DISCUSSION TITLE
Blink on Push of a Button

### DISCUSSION THREAD

**Discussion by bastian987** [arendst/Tasmota#19509] [Support request]
Hi,

I'm using a NOUS A1T and would like to achieve the following behavior:
1. if the button is pushed, the Power should toggle between on and off for a fixed number of times
2. I want to ensure the relay is never on forever, i.e., in all cases shuts the relay off after some time

To achieve 1., I have set BlinkTime and BlinkCount, if I enter 'Power 3' in the console, I get the desired behavior. To also start this behavior when the button of the NOUS A1T is pushed, I tried
Backlog ButtonTopic 0
Rule1 on Button1#state do Publish /cmnd/<my_Tasmota_device>/POWER 3 endon
Rule 1

This approach does not work yet, unfortunately. I get 
MQT: /cmnd/<my_Tasmota_device>/POWER = 3 in the console when pushing the button, but nothing happens.
Is there an obvious mistake? Is there a way to go to POWER 3 when the button is pushed without relying on MQTT (to ensure I get the desired behavior even when the MQTT server would be down)? The second part is nice to have but not required.

For 2., I would go with PulseTime <my value> to ensure the device always turns off after some time.

I would appreciate any help with 1, and feedback whether 2 is ok (in particular in this combination with using BlinkTime and BlinkCount) -- thanks for any support in advance!

**Comment by sfromis** [comment]
Do not involve MQTT when it is on the same device:
`Rule1 on Button1#state do Power blink endon`

**Reply by bastian987** [reply]
Thanks, that’s what I was looking for! One followup: now a push of the button starts the blink. Can I abort an ongoing blink (relay off) with a second push of the button if the blink is still running?

**Reply by sfromis** [reply]
You could expand on using rules to control behavior such ways. The command `Power blinkoff` exists....

**Reply by bastian987** [reply]
Thanks for the hint. I did some more reading in the docu, but was not entirely successful. What I had in mind:
`Rule1 on Button1#state do if (Power#state==3) Power blinkoff else Power blink endif endon`
However, two issues: 
- Power#state==3 is incorrect, but I couldn't find any information how to check whether blinking is active. Could you point me in the right direction?
- it requires an if statement which is not included in precompiled binaries. Any way to achieve the desired behavior using a precompiled binary?

**Reply by sfromis** [reply]
Power state is only on or off, and will cycle during blinking. Blink is not a power state but a running action. The JSON payloads you see in the console are mainly available to the usual rules, like `Power#data=Blink OFF`, and thus you can keep a var with blink state.

And you do not "require" **if**, that is merely one way of doing things, unless getting complex, things can be done as well without, one technique is to change a var, and then have a further rule trigger on that var.

If you needed power state (0/1) in a rule, you also have `%Power1%`

**Reply by bastian987** [reply]
Thanks a lot for helping me to learn, @sfromis. Highly appreciated.

I'm still wondering about the JSON payload and Power#data, unfortunately:  
"JSON payloads you see in the console" => `Power` in the terminal gives me a JSON object like this {"POWER":"OFF"}. Looking at your syntax with #data, I would have expected {"POWER":"OFF", "data": <value or list of values>}. I went through the docu now again, but didn't find info that helps me.

"thus you can keep a var with blink state": I assume from this statement that the state of the action "blink" is not stored somewhere automatically, but that I need to store a "blink=1" in a variable when blink is triggered and set it to "blink=0" somehow when the blink is completed?


"one technique is to change a var, and then have a further rule trigger on that var" => again, I'm thinking about it for quite some time now but can't figure this out :( Is there somewhere an example how to replace an if else with this technique?

Rule1 on Button1#state do var1 0 endon
Rule2 on var1==0 do <some command> endon
is equivalent to just 
Rule1 on Button1#state do <some command> endon
How does this implement an if else?

**Reply by sfromis** [reply]
Yes, you can use the [JSON-based triggers](https://tasmota.github.io/docs/Rules/#rule-trigger) to update a var for what the current blink state is. If you do than in `var1`, you could do a `var2 %var1%` upon button action, and then `on var2#state==0` and `on var2#state==1` for if/else.

And **read what you see in the console**, like
`.../POWER {"POWER":"Blink OFF"}`
That is what you can work with in rules, as I pointed out. I was not suggesting to try a command, but see what you get when blinking.

**Comment by bastian987** [comment]
Finally, I could resolve the issue, but went down another route. I want to document it, in case someone has a similar problem:
Detach relay from button with SetOption73 1, fetch the mqtt message of a button push in NodeRed and implement the logic here, send Power commands over mqtt as required.

Benefits of this path in my case:

- significantly easier to implement, change and maintain for me personally.
- I can define the wait time between blinks as a variable in home assistant and read the value of the variable in NodeRed => wait time can be adjusted in the home assistant app if needed.
- More advanced behaviors are easily possible, e.g., different timing for on and off phase, or timing that changes over the course of the loop.

Thanks again everyone for your support.
